File: Aeds/AedsII/Atividades/Trabalhos Praticos/Tp3/Num11/codigo.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;


class Jogador {

    public static class Celula {
        public Celula prox,ant;
        public Jogador elemento;

        public Celula(){
            this(null);
        }

        public Celula(Jogador elemento){
            this.elemento = elemento;
            this.ant = this.prox = null;
        }
    }

    static int tamanho=0;
    static Celula primeiro, ultimo;
    static int comparacoes=0, movimentacoes=0;
    public static Long tempo = (long) 0;

    private Integer id,altura,peso,anoNascimento;
    private String nome,universidade,cidadeNascimento,estadoNascimento;


    public String getEstadoNascimento() {
        return estadoNascimento;
    }
    public String getCidadeNascimento() {
        return cidadeNascimento;
    }
    public String getUniversidade() {
        return universidade;
    }
    public String getNome() {
        return nome;
    }
    public int getAnoNascimento() {
        return anoNascimento;
    }
    public int getPeso() {
        return peso;
    }
    public int getAltura() {
        return altura;
    }
    public int getId() {
        return id;
    }
    public void setId(int id) {
        this.id = id;
    }
    public void setAltura(int altura) {
        this.altura = altura;
    }
    public void setPeso(int peso) {
        this.peso = peso;
    }
    public void setAnoNascimento(int anoNascimento) {
        this.anoNascimento = anoNascimento;
    }
    public void setNome(String nome) {
        this.nome = nome;
    }
    public void setUniversidade(String universidade) {
        this.universidade = universidade;
    }
    public void setCidadeNascimento(String cidadeNascimento) {
        this.cidadeNascimento = cidadeNascimento;
    }
    public void setEstadoNascimento(String estadoNascimento) {
        this.estadoNascimento = estadoNascimento;
    }

    public Jogador(){
        this.id = null;
		this.nome = null;
		this.altura = null;
		this.peso = null;
		this.universidade = null;
		this.anoNascimento = null;
		this.cidadeNascimento = null;
		this.estadoNascimento = null;
    }
    public Jogador cloneJogador() {
		Jogador clone = new Jogador();
		clone.id = this.getId();
		clone.nome = this.getNome();
		clone.altura = this.getAltura();
		clone.peso = this.getPeso();
		clone.universidade = this.getUniversidade();
		clone.anoNascimento = this.getAnoNascimento();
		clone.cidadeNascimento = this.getCidadeNascimento();
		clone.estadoNascimento = this.getEstadoNascimento();
		return clone;
	}

    /*----------------------------------------------------------------------------------------------------------------------------------------- INSERCAO DE DADOS INICIO */

    public static void inserirElementos(){
        String input = MyIO.readLine();
        while (!input.equals("FIM")) {
            try {
                Integer id = Integer.parseInt(input);
                Jogador jogador = new Jogador();
                jogador.SetaId(id);
                setaCelula(jogador);

            } catch (Exception e) {}
            input = MyIO.readLine();
        } 
    }

    public static void setaCelula(Jogador jogador){
        Celula celula = new Celula(jogador);
        ultimo.prox = celula;
        ultimo.prox.ant = ultimo;
        ultimo = ultimo.prox;
    }

    public void SetaId(int x) throws Exception {
        FileReader file = new FileReader("/tmp/players.csv"); 
        BufferedReader buffer = new BufferedReader(file);
        String linha;

        while ((linha = buffer.readLine()) != null) {
            String numId = linha.substring(0, linha.indexOf(","));

            try{
                if(Integer.parseInt(numId) == x) {
                    this.setaDados(linha);
                    //this.mostraDados();
                    break;
                }
            }catch(NumberFormatException e){}
        }
        buffer.close();
        file.close();
    }

    public void setaDados(String linha){
        String[] parte = linha.split(",");

        this.setId(Integer.parseInt(parte[0]));
        this.setNome(parte[1]);
        this.setAltura(Integer.parseInt(parte[2]));
        this.setPeso(Integer.parseInt(parte[3]));
        if(parte.length>4 && !parte[4].isEmpty()) this.setUniversidade(parte[4]); else this.setUniversidade("nao informado");
        this.setAnoNascimento(Integer.parseInt(parte[5]));
        if (parte.length>6 && !parte[6].isEmpty()) this.setCidadeNascimento(parte[6]); else this.setCidadeNascimento("nao informado");
        if (parte.length>7 && !parte[7].isEmpty()) this.setEstadoNascimento(parte[7]); else this.setEstadoNascimento("nao informado");

    }

    public void mostraDados(){
        MyIO.println("[" + this.getId() + " ## " + this.getNome() + " ## " + this.getAltura() + " ## " + this.getPeso() + " ## " + this.getAnoNascimento() + " ## " 
        + this.getUniversidade() + " ## " + this.getCidadeNascimento() + " ## " + this.getEstadoNascimento() + "]");
    }

    public static void MostraLista() {
		for(Celula i=primeiro.prox; i!=null; i=i.prox) {           /* SE EU COLOCO ATE I.PROX!=NULL ELE VAI ANDAR 1 ELEMENTO NA FRENTE */
            i.elemento.mostraDados();
		}
	}

    /*----------------------------------------------------------------------------------------------------------------------------------- FIM INSERCAO DE DADOS */
    /* ------------------------------------------------------------------------------------------------------------------------------------------- INICIO ORDENAÇÃO */
    
    public static void ordenaElementos(){
        quicksort(primeiro.prox,ultimo);
        OrgNome(tamanho);
    }
    
    public static void quicksort(Celula esq, Celula dir){
        Celula pivo=esq, i=esq.prox, j=dir;

        while(j.prox!=i && j.prox!=i.ant){
            while(i!=null && i.elemento.estadoNascimento.compareTo(pivo.elemento.estadoNascimento)<0){ i=i.prox; comparacoes++; }
            while(j.elemento.estadoNascimento.compareTo(pivo.elemento.estadoNascimento)>0){ j=j.ant; comparacoes++; }
            if (j.prox!=i){
                swap(i, j);
                if(i.prox!=null) i=i.prox;
                if(j.ant!=null) j=j.ant;
            }
        }
        swap(pivo, j);

        if (j!=esq) quicksort(esq, j.ant);
        if (j!=dir) quicksort(j.prox, dir);
    }   
    
    public static void OrgNome(int tam) {
        for (Celula i = primeiro.prox; i.prox != null; i = i.prox) {
            Celula menor = i;
            for (Celula j = i.prox; j != null; j = j.prox) {
                if (i.elemento.estadoNascimento.equals(j.elemento.estadoNascimento)) {
                    if (menor.elemento.nome.compareTo(j.elemento.nome) > 0) {
                        menor = j;
                    }
                }
            }
            swap(menor, i);
        }
    }

    public static void swap(Celula i, Celula j){
        Jogador tmp =  i.elemento;
        i.elemento = j.elemento;
        j.elemento = tmp;
        movimentacoes+=3;
    }

    /* ------------------------------------------------------------------------------------------------------------------------------------------- FIM ORDENAÇÃO */

	public static void ArqLog(long tempo) {
		String nomeArq = "matrícula_insercao.txt";
		String mtr = "1441272";
		
		try {
            File arq = new File(nomeArq);
            arq.createNewFile();
        	try {
    			FileWriter file = new FileWriter(nomeArq, false); 
    			BufferedWriter buffer = new BufferedWriter(file);

    			buffer.write("Matricula:" + mtr + '\t' + "Tempo de Execução: " + tempo/1000 + "ms" + '\t' + 
                "Numero de Comparações: " + comparacoes + '\t' + '\t' + "Numero de Movimentações: " + movimentacoes);
    			buffer.close();

    		}catch(IOException e) {e.printStackTrace();}
        }catch(IOException e){e.printStackTrace();}
	}


    public static void main(String[] args) {                    // Compilar:    javac .\Jogador.java        Executar:    type .\pub.in | java Jogador > result.txt
        primeiro = ultimo = new Celula();                       // Compilar:    javac Jogador.java          Executar:    java Jogador < pub.in > result.txt      
        inserirElementos();  
        
        long start = System.nanoTime();
        ordenaElementos();
        long fim = System.nanoTime();
        tempo = fim-start;
        
        MostraLista();
        ArqLog(tempo);
    }
}

File: Aeds/AedsII/Atividades/Trabalhos Praticos/Tp3/Num 8/codigo.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>
#include <time.h>

/*
    Dev: Tulio Gomes Braga
    Matricula: 1441272
    Curso: CC-PucMinas - 14/11/2023
*/

typedef struct {
    int id;
    int altura;
    int peso;
    int anoNascimento;
    char nome[100];
    char universidade[100];
    char cidadeNascimento[100];
    char estadoNascimento[100];

} Jogador;

typedef struct Celula {
    Jogador elemento;
    struct Celula* prox;
    struct Celula* ant; 

}Celula;

Celula* newCelula(Jogador elemento){
    Celula* nova = (Celula*) malloc(sizeof(Celula));
    nova->elemento = elemento;
    nova->prox = NULL;
    nova->ant=NULL;

    return nova;    
}


Celula* primeiro; Celula* ultimo;
int tamanho = 0, movimentacoes, comparacoes;
clock_t start,end;

void inserirElementos();
void setaId(int id);
void parteDados(const char *linha);
void setaDados(char *linha);
void mostraDados(Jogador jogador);
void mostrarTodos();
void orgElementos();
void quicksort(Celula* esq, Celula* dir);
void orgNome(int tam);
void swap(Celula* i, Celula* j);
void ArqLog();


void inserirElementos(){
    char input[100];
    scanf("%s", input);

    while(strcmp(input, "FIM")!= 0){
        int id = atoi(input);
        setaId(id);
        scanf("%s", input);
    }
}

void setaId(int id) {
    FILE *file;
    file = fopen("/tmp/players.csv", "r");

    char linha[100];
    char resp[100];
    bool achou = false;
    char *token;

    fgets(linha, sizeof(linha), file);

    while (fgets(linha, sizeof(linha), file) != NULL && !achou){
        strcpy(resp,linha);
        token = strtok(linha,",");
        int num = atoi(token);

        if(id==num){
            parteDados(resp);
            achou=true;
        }
        token=strtok(NULL,",");

    }
    fclose(file);
}

void parteDados(const char *x) {

    int len = strlen(x);
    char linha[len];
    strcpy(linha,x);

    linha[len-1]='\0';
    int j = 0;
    char *nvlinha = malloc(len + 4);


    for (int i = 0; i < len; i++) {
        if (linha[i]==',' && linha[i+1]==',') {
            nvlinha[j]=linha[i];
            nvlinha[j+1]=' ';
            nvlinha[j+2]=',';
            j+=3;
            i++;
        } else {
            nvlinha[j] = linha[i];
            j++;
        }
    }

    if(nvlinha[j-2]==','){
        nvlinha[j-1]=' ';
    }
    nvlinha[j]='\0';

    setaDados(nvlinha);
    free(nvlinha);
}

void setaDados(char *linha){
    Jogador jogador;

    int t=0;
    char *parte[8];
    char *token = strtok(linha, ",");

    while (token != NULL && t < 8) {
        parte[t] = token;
        token = strtok(NULL, ",");          //MANTEM UM ESTADO INTERNO PARA RASTREAR A POSICAO ATUAL NA STRING
        t++;
    }

    jogador.id = atoi(parte[0]);
    if(strcmp(parte[1], " ")== 0)strcpy(jogador.nome, "nao informado");else strcpy(jogador.nome, parte[1]);
    jogador.altura = atoi(parte[2]);
    jogador.peso = atoi(parte[3]);
    if(strcmp(parte[4], " ")== 0)strcpy(jogador.universidade, "nao informado");else strcpy(jogador.universidade, parte[4]);
    jogador.anoNascimento = atoi(parte[5]);
    if(strcmp(parte[6], " ")== 0)strcpy(jogador.cidadeNascimento, "nao informado");else strcpy(jogador.cidadeNascimento, parte[6]);
    if(strcmp(parte[7], " ")== 0)strcpy(jogador.estadoNascimento, "nao informado");else strcpy(jogador.estadoNascimento, parte[7]);

    //mostraDados(jogador);
    if(tamanho!=0){
        ultimo->prox = newCelula(jogador);
        ultimo->prox->ant = ultimo;
        ultimo = ultimo->prox;
        tamanho++;
    }else{
        primeiro = newCelula(jogador);
        ultimo = primeiro;
        tamanho++;
    }
}

void mostraDados(Jogador jogador){
    printf("[%d ## %s ## %d ## %d ## %d ## %s ## %s ## %s]\n",
        jogador.id,
        jogador.nome,   
        jogador.altura,
        jogador.peso,
        jogador.anoNascimento,
        jogador.universidade,
        jogador.cidadeNascimento,
        jogador.estadoNascimento    
    );
}

void mostrarTodos(){
    Celula* i;
    for(i=primeiro; i!=NULL; i=i->prox){
        mostraDados(i->elemento);
    }
}

void orgElementos(){
    quicksort(primeiro,ultimo);
    orgNome(tamanho);
}

void quicksort(Celula* esq, Celula* dir) {
    if (esq != NULL && dir != NULL && esq != dir && dir->prox != esq) {
        Celula* pivo = esq;
        Celula* i = esq->prox;
        Celula* j = dir;

        while (j->prox != i && j->prox != i->ant) {
            while (i != NULL && strcmp(i->elemento.estadoNascimento, pivo->elemento.estadoNascimento) < 0) {
                i = i->prox;
                comparacoes++;
            }
            while (strcmp(j->elemento.estadoNascimento, pivo->elemento.estadoNascimento) > 0) {
                j = j->ant;
                comparacoes++;
            }
            if (j->prox != i){
                swap(i, j);
                if (i->prox != NULL) i = i->prox;
                if (j->ant != NULL) j = j->ant;
            }
        }
        swap(pivo, j);

        if (j != esq) quicksort(esq, j->ant);
        if (j != dir) quicksort(j->prox, dir);
    }
}

void orgNome(int tam) {
    for (Celula* i = primeiro; i->prox != NULL; i = i->prox) {
        Celula* menor = i;
        for (Celula* j = i->prox; j != NULL; j = j->prox) {
            if (strcmp(i->elemento.estadoNascimento, j->elemento.estadoNascimento) == 0) {
                comparacoes++;
                if (strcmp(menor->elemento.nome, j->elemento.nome) > 0) {
                    menor = j;
                }
            }
        }
        swap(menor, i);
    }
}

void swap(Celula* i, Celula* j) {
    Jogador tmp = i->elemento;
    i->elemento = j->elemento;
    j->elemento = tmp;
    movimentacoes += 3; 
}

void ArqLog(){
    int matricula = 1441272;
    float tempo;
    tempo = (float)(end-start)*1000;

    FILE *file = fopen("matrícula_quicksort.txt","w");
    fprintf(file,"Matricula: %i\tTempo: %f\tComparacoes: %i\tMovimentacoes: %i",matricula, tempo, comparacoes, movimentacoes);
}


int main(){ 
    inserirElementos();
    orgElementos();
    mostrarTodos();
    ArqLog();

    return 0;
}